src/lib/server/xlat_func.c

```diff
@@ -1439,32 +1439,48 @@ static xlat_action_t base64_xlat(TALLOC_CTX *ctx, fr_cursor_t *out,
 	return XLAT_ACTION_DONE;
 }
 
-/** Convert base64 to hex
+/** Decode base64 string
  *
- * Example: "%{base64tohex:Zm9v}" == "666f6f"
+ * Example: "%{base64decode:Zm9v}" == "foo"
  */
-static ssize_t base64_to_hex_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-				  UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-				  REQUEST *request, char const *fmt)
+static xlat_action_t xlat_base64_decode(TALLOC_CTX *ctx, fr_cursor_t *out,
+					REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+					fr_value_box_t **in)
 {
-	uint8_t decbuf[1024];
+	size_t		alen;
+	ssize_t		declen;
+	uint8_t		*decbuf;
+	fr_value_box_t	*vb;
+
+	/*
+	 *	If there's no input, there's no output
+	 */
+	if (!in) return XLAT_ACTION_DONE;
+
+	if (fr_value_box_list_concat(ctx, *in, in, FR_TYPE_OCTETS, true) < 0) {
+		RPEDEBUG("Failed concatenating input");
+		return XLAT_ACTION_FAIL;
+	}
+
+	alen = FR_BASE64_DEC_LENGTH((*in)->vb_length);
 
-	ssize_t declen;
-	ssize_t len = strlen(fmt);
+	MEM(decbuf = talloc_array(ctx, uint8_t, alen));
 
-	declen = fr_base64_decode(decbuf, sizeof(decbuf), fmt, len);
+	declen = fr_base64_decode(decbuf, alen, (*in)->vb_strvalue, (*in)->vb_length);
 	if (declen < 0) {
+		talloc_free(decbuf);
 		REDEBUG("Base64 string invalid");
-		return -1;
+		return XLAT_ACTION_FAIL;
 	}
 
-	if ((size_t)((declen * 2) + 1) > outlen) {
-		REDEBUG("Base64 conversion failed, output buffer exhausted, needed %zd bytes, have %zd bytes",
-			(declen * 2) + 1, outlen);
-		return -1;
-	}
+	MEM(vb = fr_value_box_alloc_null(ctx));
+
+	fr_value_box_memdup(vb, vb, NULL, decbuf, declen, false);
```

Comment:
Use fr_value_box_memsteal instead to avoid the extra alloc and memcpy

Reply:
No, that's not possible as the buffer for decoding the base64 data can be longer that the actual decoded data. In some cases, we would end up with too much data. We need to only copy `declen` bytes.